core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java

```diff
@@ -5,45 +5,59 @@
 
 package org.opensearch.sql.planner.optimizer.rule;
 
-import static org.opensearch.sql.planner.optimizer.pattern.Patterns.table;
-
-import com.facebook.presto.matching.Capture;
 import com.facebook.presto.matching.Captures;
 import com.facebook.presto.matching.Pattern;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
 import lombok.Getter;
 import lombok.experimental.Accessors;
+import org.opensearch.sql.planner.logical.LogicalPaginate;
 import org.opensearch.sql.planner.logical.LogicalPlan;
 import org.opensearch.sql.planner.logical.LogicalRelation;
 import org.opensearch.sql.planner.optimizer.Rule;
-import org.opensearch.sql.storage.Table;
-import org.opensearch.sql.storage.read.TableScanBuilder;
+import org.opensearch.sql.planner.physical.PhysicalPlan;
 
 /**
  * Rule to create a paged TableScanBuilder in pagination request.
  */
-public class CreatePagingTableScanBuilder implements Rule<LogicalRelation> {
-  /** Capture the table inside matched logical relation operator. */
-  private final Capture<Table> capture;
-
+public class CreatePagingTableScanBuilder implements Rule<LogicalPaginate> {
+  /** Capture the table inside matched logical paginate operator. */
+  private LogicalPlan relationParent = null;
   /** Pattern that matches logical relation operator. */
   @Accessors(fluent = true)
   @Getter
-  private final Pattern<LogicalRelation> pattern;
+  private final Pattern<LogicalPaginate> pattern;
 
   /**
    * Constructor.
    */
   public CreatePagingTableScanBuilder() {
-    this.capture = Capture.newCapture();
-    this.pattern = Pattern.typeOf(LogicalRelation.class)
-        .with(table().capturedAs(capture));
+    this.pattern = Pattern.typeOf(LogicalPaginate.class).matching(lp -> {
+      Deque<LogicalPlan> plans = new ArrayDeque<>();
+      plans.push(lp);
+      do {
+        var plan = plans.pop();
+        if (plan.getChild().stream().anyMatch(LogicalRelation.class::isInstance)) {
+          if (plan.getChild().size() > 1) {
+            throw new UnsupportedOperationException();
+          }
+          relationParent = plan;
+          return true;
+        }
+        plan.getChild().forEach(plans::push);
+      } while (!plans.isEmpty());
+      return false;
+    });
   }
 
+
   @Override
-  public LogicalPlan apply(LogicalRelation plan, Captures captures) {
-    TableScanBuilder scanBuilder = captures.get(capture)
-        .createPagedScanBuilder(plan.getPageSize());
-    // TODO: Remove this after Prometheus refactored to new table scan builder too
-    return (scanBuilder == null) ? plan : scanBuilder;
+  public LogicalPlan apply(LogicalPaginate plan, Captures captures) {
+    var logicalRelation = (LogicalRelation) relationParent.getChild().get(0);
+    var scan = logicalRelation.getTable().createPagedScanBuilder(logicalRelation.getPageSize());
```

Comment:
If we know (and set) page size here, we can omit `PushPageSize` then.

Reply:
True. Would this make your push size PR unnecessary?

Reply:
Yes, and good if so.